From a36030eb22d9d27ecb68586fa848c9a371436460 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Sat, 15 Feb 2020 23:28:57 +0200
Subject: [PATCH] [SW-8229] goya2: fix hw handeling of pcie resets

Disable control of hot reset via the hardware, as this
caused a issue with unstable link losing the link completely
because of hw miss handeling the case.

Enable control of FLR reset vi hardware.

Change-Id: I275b842b1ce4fe2fe894918c0237cbd428c5f27d
Signed-off-by: Noam Lessner <nlessner@habana.ai>
---

diff --git a/common/pcie.c b/common/pcie.c
index 2ef5750..ded39a4 100644
--- a/common/pcie.c
+++ b/common/pcie.c
@@ -25,7 +25,7 @@
 static void pcie_config(void)
 {
 	uint8_t sram_init_done;
-	uint32_t boot_strap;
+	uint32_t boot_strap, flr_ctrl;
 	uint8_t ltssm_en;
 
 	pcie_loop_while_prstn(PCIE_PRSTN_ASSERT);
@@ -67,6 +67,16 @@
 		regmap_writel(0x7F, IATU_UPPER_TARGET_ADDR_OFF_INBOUND);
 		regmap_writel(0, IATU_REGION_CTRL_1_OFF_INBOUND);
 		regmap_writel(0xC0080000, IATU_REGION_CTRL_2_OFF_INBOUND);
+
+		/* enable flr handling by HW state machine */
+		flr_ctrl = regmap_readl(mmPCIE_AUX_FLR_CTRL);
+		flr_ctrl |= (1 << PCIE_AUX_FLR_CTRL_FSM_EN_SHIFT);
+		regmap_writel(flr_ctrl, mmPCIE_AUX_FLR_CTRL);
+
+		/* disable HOT reset handling by HW state machine */
+		regmap_writel(PCIE_WRAP_CPU_HOT_RST_CTRL_MASK,
+			      mmPCIE_WRAP_CPU_HOT_RST);
+
 		regmap_writel(0x3, mmPCIE_AUX_LTSSM_EN);
 	}
 
